File: coding/src/com/algos/core/li10_tree/bst/revised/L173_BinarySearchTreeIterator.java
package  com.algos.core.li10_tree.bst.revised;

import java.util.Stack;

import  com.algos.core.li30_model.TreeNode;

class L173_BinarySearchTreeIterator {

    public static void main(String[] args) {
        //Bst_L173_BinarySearchTreeIterator obj = new Bst_L173_BinarySearchTreeIterator();

    }
    // Problem: https://leetcode.com/problems/binary-search-tree-iterator/
    // Idea: The first processed element would be left most. 
    // Hence, push all the elements to stack until the left most so that left most is available on top of the stack.
    // While process left most, it won't have any further lefts. Hence, the next processed will be left most in it right child. Push all the elements till left most in its right child.
    // Time: O(logn)
    // space: O(logn)
    Stack<TreeNode> stack = new Stack<TreeNode>();
    public L173_BinarySearchTreeIterator(TreeNode root) {
        TreeNode curr = root;
        while(curr != null){
            stack.push(curr);
            curr = curr.left;
        } 
    }
    
    public int next() {
        TreeNode next = stack.pop();
        TreeNode curr = next.right;
        while(curr != null){
            stack.push(curr);
            curr = curr.left;
        }
        return next.val;
    }
    
    public boolean hasNext() {
        return !stack.isEmpty();
    }
}

File: dsa/topics/10_trees/bst/python/l173__binary_search_tree_iterator.py
# Tree / Stack
# Problem: https://leetcode.com/problems/binary-search-tree-iterator/
# Idea:
# 1. The first processed element is the leftmost node.
# 2. Push all left children to stack so that leftmost is on top.
# 3. When processing a node, if it has a right child, push all its left children to stack.
# Time: O(log n) amortized
# Space: O(log n)

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BSTIterator:
    def __init__(self, root):
        self.stack = []
        curr = root
        while curr:
            self.stack.append(curr)
            curr = curr.left

    def next(self):
        next_node = self.stack.pop()
        curr = next_node.right
        while curr:
            self.stack.append(curr)
            curr = curr.left
        return next_node.val

    def hasNext(self):
        return len(self.stack) > 0


# -------- Test --------
# Construct BST: 
#        7
#       / \
#      3   15
#          / \
#         9  20
root = TreeNode(7)
root.left = TreeNode(3)
root.right = TreeNode(15, TreeNode(9), TreeNode(20))

iterator = BSTIterator(root)
result = []
while iterator.hasNext():
    result.append(iterator.next())
print(result)  # Expected: [3, 7, 9, 15, 20]
